tests/core/test_optimization_variable.py

```diff
@@ -0,0 +1,364 @@
+        assert variable.data == {}
+        assert variable.levels == []
+        assert variable.marginals == []
+        assert variable.columns is not None
+        assert variable.columns[0].name == indexset.name
+        assert variable.constrained_to_indexsets == [indexset.name]
+
+        with pytest.raises(OptimizationVariable.NotFound):
+            _ = run.optimization.variables.get("Variable 2")
+
+    def test_variable_add_data(self, platform: ixmp4.Platform):
+        run = platform.runs.create("Model", "Scenario")
+        indexset, indexset_2 = tuple(
+            IndexSet(_backend=platform.backend, _model=model)
+            for model in create_indexsets_for_run(platform=platform, run_id=run.id)
+        )
+        indexset.add(elements=["foo", "bar", ""])
+        indexset_2.add(elements=[1, 2, 3])
+        # pandas can only convert dicts to dataframes if the values are lists
+        # or if index is given. But maybe using read_json instead of from_dict
+        # can remedy this. Or maybe we want to catch the resulting
+        # "ValueError: If using all scalar values, you must pass an index" and
+        # reraise a custom informative error?
+        test_data_1 = {
+            indexset.name: ["foo"],
+            indexset_2.name: [1],
+            "levels": [3.14],
+            "marginals": [0.000314],
+        }
+        variable = run.optimization.variables.create(
+            "Variable",
+            constrained_to_indexsets=[indexset.name, indexset_2.name],
+        )
+        variable.add(data=test_data_1)
+        assert variable.data == test_data_1
+        assert variable.levels == test_data_1["levels"]
+        assert variable.marginals == test_data_1["marginals"]
+
+        variable_2 = run.optimization.variables.create(
+            name="Variable 2",
+            constrained_to_indexsets=[indexset.name, indexset_2.name],
+        )
+
+        with pytest.raises(
+            AssertionError, match=r"must include the column\(s\): marginals!"
+        ):
+            variable_2.add(
+                pd.DataFrame(
+                    {
+                        indexset.name: [None],
+                        indexset_2.name: [2],
+                        "levels": [1],
+                    }
+                ),
+            )
+
+        with pytest.raises(
+            AssertionError, match=r"must include the column\(s\): levels!"
+        ):
+            variable_2.add(
+                data=pd.DataFrame(
+                    {
+                        indexset.name: [None],
+                        indexset_2.name: [2],
+                        "marginals": [0],
+                    }
+                ),
+            )
+
+        # By converting data to pd.DataFrame, we automatically enforce equal length
+        # of new columns, raises All arrays must be of the same length otherwise:
+        with pytest.raises(ValueError, match="All arrays must be of the same length"):
```

Comment:
ditto like with the error from above! not sure if something like this occurs somewhere else as well?